pkg/applicationserver/grpc_deviceregistry.go

```diff
@@ -215,6 +216,9 @@ func (r asEndDeviceRegistryServer) Delete(ctx context.Context, ids *ttnpb.EndDev
 	}
 	var evt events.Event
 	_, err := r.AS.deviceRegistry.Set(ctx, *ids, nil, func(dev *ttnpb.EndDevice) (*ttnpb.EndDevice, []string, error) {
+		if dev == nil || dev.ApplicationIdentifiers != ids.ApplicationIdentifiers {
```

Comment:
```suggestion
		if dev == nil {
```
`dev.ApplicationIdentifiers != ids.ApplicationIdentifiers` means a bug in the registry, at which point we should just `panic` instead. In any case I don't think we should do this, if you really insist on having this check - please refactor it into a panic.